control/autopilot/include/autopilot/autopilot_inl.h

```diff
@@ -857,7 +878,7 @@ quadrotor_common::ControlCommand AutoPilot::start(
   }
 
   if (timeInCurrentState() > optitrack_start_land_timeout_
-      || reference_state_.position.z() >= optitrack_start_height_)
+      || state_estimate.position.z() >= optitrack_start_height_)
```

Comment:
why the change to `state_estimate` from `reference_state` ?

Reply:
Whats the intention behind checking what the reference height is? Doesn't it make more sense to check what your actual height is and cancel the take-off once you reached this height?

Reply:
if you look further in the function the reference is moved up and up as the loop is repeated. I would say that using the old method makes it more robust and smoother but this is only a feeling. By that I mean, hover WILL be reached even if you have an offset for idk which reason.

Reply:
So ideally the drone does what the reference tells it to, and in that case it does not make much of a difference. However, if e.g. the drone has a wrong thrust mapping and flies much lower than the reference, when using the state estimate, it will go up until it reached the start height (with a reference much higher) but then when it switches back to hover the reference is set to the current state estimate and the drone will drop by that much again which also does not result in the desired height. So by checking the reference height at least we get a smooth behavior in this case. Anyhow, if the drone does not follow the reference well it needs to be fixed...
Bottom line: I would change this back to reference state

Reply:
There you go, the reason I was searching for! Make sense to me

Reply:
OK, clarification:
When using the MPC, it lags behind since your take-off maneuver has a rising position, fixed velocity but zero acceleration and hover thrust, which is unfeasible at the beginning of the start maneuver.
This is no problem for feasible control with the MPC, but it means that the it lags behind due to the compromise of positon-, velocity-, acceleration- and thrust-error.

Possible solution:
1. Ignore ~0.15m z-offset after start.
2. Add a takeoff acceleration (which could be a parameter) at the beginning of the start maneuver.